### PR TITLE
adding Analyticsv3Sample

### PR DESCRIPTION
**Pull request by LindaLawton**
Code review welcome in the Analyticsv3Sample project

### CODE REVIEW

File: Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        }
+
+
+    }
+
+    public static class extensions
+    {
+
+        public static GaData getNextRows(this GaData data, AnalyticsService service)
```
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs]: This isnt genrated i am working with the client library team on this.  [pagination](https://github.com/google/google-api-dotnet-client/issues/632)

File: Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        /// <param name="startdate">Start date for fetching Analytics data. Requests can specify a start date formatted as YYYY-MM-DD, or as a relative date (e.g., today, yesterday, or 7daysAgo). The default value is 7daysAgo.</param>
+        /// <param name="enddate">End date for fetching Analytics data. Request can should specify an end date formatted as YYYY-MM-DD, or as a relative date (e.g., today, yesterday, or 7daysAgo). The default value is yesterday.</param>
+        /// <param name="metrics">A comma-separated list of Analytics metrics. E.g., 'ga:sessions,ga:pageviews'. At least one metric must be specified.</param>
+        /// <returns></returns>
+        public static GaData get(AnalyticsService service, string ids, string startdate, string enddate, string metrics)
+        {
+            try
+            {
+                var request = service.Data.Ga.Get(ids, startdate, enddate, metrics);
+                request.Dimensions = "ga:hour";
```
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs]: optional items aren't generated.  I was thinking of maybe having an over load of the methods with optional values.

**Comment by LindaLawton** [comment]
Question: When i upload all of the generated samples what do i do with the old samples?  They are technically quick examples used to answer questions.   These samples will be standardized.

File: Analyticsv3Sample/AnalyticsSampleProject/Program.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    /// In order to run this you must install the nuget package.  Please consult the readme.
+    /// 
+    /// Link to Nuget Package: https://www.nuget.org/packages/Google.Apis.analytics.v3/
+    /// PM> Install-Package Google.Apis.Analytics.v3 
+    /// </summary>
+    class Program
+    {
+        static void Main(string[] args)
+        {
+
```
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/Program.cs]: - What should we include here?  
- Should we give an example of authentication?   (OAuth2, service account, API Key or all three)
- Should we give an example of using the methods and looping though the results?

File: Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs (hunk 2 above)
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs]: You can provide another method that returns the request itself, then the user can add another optional paramters that he likes to add.
Since after that he expected to just call Execute or ExecuteAsync you don't need to provide another method that just calls it.

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        /// Documentation https://developers.google.com/accounts/docs/OAuth2
+        /// </summary>
+        /// <param name="clientId">From Google Developer console https://console.developers.google.com</param>
+        /// <param name="clientSecret">From Google Developer console https://console.developers.google.com</param>
+        /// <param name="userName">The user to authorize.</param>
+        /// <returns>a valid AnalyticsService</returns>
+        public static AnalyticsService AuthenticateOauth(string clientId, string clientSecret, string userName)
+        {
+            if (string.IsNullOrEmpty(clientId))
+                throw new Exception("clientId is required.");
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:112]: Consider changing your exception to Argument Exception (https://msdn.microsoft.com/en-us/library/system.argumentexception(loband).aspx)

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Analytics.v3;
+using Google.Apis.Services;
+using Google.Apis.Util.Store;
+using System;
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:23]: [optional]
System first and then Google usings

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+	 /// <summary>
+        /// The Google APIs Client Library for .net uses the client_secrets.json file format for storing the client_id, client_secret, and other OAuth 2.0 parameters.
+        /// The client_secrets.json file format is a JSON formatted file containing the client ID, client secret, and other OAuth 2.0 parameters.
+        ///      
+        /// This file can be obtained from Google Developers console:   https://console.developers.google.com/project?authuser=0
+        /// </summary>
+        /// <param name="clientSecretJson">Path to the client secret json file from Google Developers console.</param>
+        /// <param name="userName">The user to authorize.</param>
+        /// <returns>a valid AnalyticsService</returns>
+        public static AnalyticsService AuthenticateOauth(string clientSecretJson, string userName)
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:54]: Do you think to create an async version, and rename to: GetAuthenticatedServiceAsync?

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                    Console.WriteLine("Credential file saved to: " + credPath);
+                }
+
+                // Create Analytics API service.
+                var service = new AnalyticsService(new BaseClientService.Initializer()
+                    {
+                        HttpClientInitializer = credential,
+                        ApplicationName = "Analytics Authentication Sample",
+                    });
+                return service;
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:91]: You can just return the service without have a variable for it, I would change the code to:
return new AnalyticsService(new BaseClientService.Initializer
                    {
                        HttpClientInitializer = credential,
                        ApplicationName = "Analytics Authentication Sample",
                    });

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        }
+
+        /// <summary>
+        /// Authenticating to Google using a Service account
+        /// Documentation: https://developers.google.com/accounts/docs/OAuth2#serviceaccount
+        /// </summary>
+        /// <param name="serviceAccountEmail">From Google Developer console https://console.developers.google.com</param>
+        /// <param name="keyFilePath">Location of the Service account key file downloaded from Google Developer console https://console.developers.google.com</param>
+        /// <returns>a valid AnalyticsService</returns>
+        public static AnalyticsService AuthenticateServiceAccount(string serviceAccountEmail, string keyFilePath)
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:161]: I don't think that all services support ServiceAccount, how do you know if you need to generate it or not. Do you just generate it for every service.
If so, add some context that says that the service itself might not support service account...

Makes sense?

File: Analyticsv3Sample/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
+
+## DAIMTO Google Analytics API Sample for .NET  ##
+
+View and manage your Google Analytics data
+
+## Description ##
+The DAIMTO Google Analytics API Sample for .NET instructs you in the usage of the Google APIs client library for .NET.
+
+The sample contains OAuth2.0 authentication, service account, and and public API.   As well as methods for each of the Resources avaliable in the Analytics API.
```
Comment (peleyal) [inline Analyticsv3Sample/README.md:9]: Consider adding space between OAuth to 2.0

Maybe also rephrase it to:
The sample contains usage of the Google Analytics V3 using OAuth2.0 and Service Account.

File: Analyticsv3Sample/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+* [Nuget package Google Apis Analytics v3](https://www.nuget.org/packages/Google.Apis.analytics.v3/)
+  *   PM> Install-Package Google.Apis.Analytics.v3
+
+SETUP
+=================================
+
+Make sure you create your own Oauth credentials at https://console.developers.google.com/.
+
+Walk though on how to create the diffrent credentails can be found:
+* [Google Developer console service account](http://www.daimto.com/google-developer-console-service-account/)
```
Comment (peleyal) [inline Analyticsv3Sample/README.md:26]: Just something you want to notice.
The Developers Console API was changed lately, make sure that the tutorial are still up to date.

File: Analyticsv3Sample/README.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+Source Code
+=================================
+
+The source code for the Google APIs .Net client library can be found at [Google APIs Client Library for .NET](https://github.com/google/google-api-dotnet-client)
+The source code for the Google Analytics API csharp can be found at [Analytics v3 csharp API](https://developers.google.com/resources/api-libraries/download/analytics/v3/csharp).
+
+Support Forums
+=================================
+If you find an issue with in the sample project I would appricate you adding an issue in the [issue tracker](https://github.com/LindaLawton/Google-Dotnet-Samples).  
+Please supply the name of the sample you are repporting an issue for in the title. Example: \[Analytics Sample\] Doesnt work.
```
Comment (peleyal) [inline Analyticsv3Sample/README.md:48]: Maybe change to:

Please include the name of the sample in the issue title. Example...

**Comment by peleyal** [comment]
I added several comments.
Well done! Good job :+1:

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 8 above)
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:161]: I just generate service account and Oauth2 methods for all the discovery APIs that have Auth.     
- Question: Can detect in discovery service if it supports service account or not?
  I know YouTube and blogger don't,  let me look into it.  

We could make a list of the ones we know don't support it and don't build for those?  I think i would rather do that then add a note.  Users have a bad habit of not reading notes and just doing it anyway.

File: Analyticsv3Sample/README.md (hunk 10 above)
Comment (LindaLawton) [inline Analyticsv3Sample/README.md:26]: People normally add a comment or email me as soon as something breaks.  I try and keep things up to date.   I do need to update the pictures on my begging developing with google series.   I have just been waiting for them to stop changing it.   They have updated a few things this week probably fixing bugs.

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 5 above)
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:23]: Fixed Any special reason why?

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 6 above)
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:54]: I have just added the methods i use most often.   We can create what ever we like.   
-  Lets start thinking about best practice as well.  
-  Do you think we should add a another with GetAuthenticatedServiceAsync.  
-  What will the users need? 
-  What do we want to show them?

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 7 above)
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:91]: Fixed and good point thanks

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 4 above)
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:112]: I will add this in the next version these were not generated it was apparently something i was playing with.

File: Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+using System.Linq;
+using System.Text;
+using Google.Apis.Analytics.v3;
+using Google.Apis.Analytics.v3.Data;
+
+namespace DAIMTO.Analytics.v3.Sample
+{
+    class DataGaSample
+    {
+		/// <summary>
```
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs:34]: I fixed the spacing on this.  However for some reason when its uploaded to github it messes it up again.  On my PC the spacing is correct.   Need to figure out what git is doing.

File: Analyticsv3Sample/AnalyticsSampleProject/AnalyticsSampleProject.csproj [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml.Linq" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Xml" />
+  </ItemGroup>
+  <ItemGroup> 
+    <Compile Include="Program.cs" />
+	<Compile Include="DataGaSample.cs" />
```
Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/AnalyticsSampleProject.csproj:45]: I keep forgetting to add StandardGoogleAuthSample.cs file is in the folder but not include in the project until we add it here.

File: Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+		/// <param name="ids">Unique table ID for retrieving Analytics data. Table ID is of the form ga:XXXX, where XXXX is the Analytics view (profile) ID.</param>
+		/// <param name="startdate">Start date for fetching Analytics data. Requests can specify a start date formatted as YYYY-MM-DD, or as a relative date (e.g., today, yesterday, or 7daysAgo). The default value is 7daysAgo.</param>
+		/// <param name="enddate">End date for fetching Analytics data. Request can should specify an end date formatted as YYYY-MM-DD, or as a relative date (e.g., today, yesterday, or 7daysAgo). The default value is yesterday.</param>
+		/// <param name="metrics">A comma-separated list of Analytics metrics. E.g., 'ga:sessions,ga:pageviews'. At least one metric must be specified.</param>
+		/// <returns>GaData </returns>
+	    public static GaData Get(AnalyticsService service, string ids, string startdate, string enddate, string metrics)
+	    {
+		    //Note Genrate Argument Exception (https://msdn.microsoft.com/en-us/library/system.argumentexception(loband).aspx)
+		    try
+            {
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/DataGaSample.cs:48]: Please fix the alignment (also on line 35, and other places in the file)

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs (hunk 8 above)
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:161]: I'm not aware of any way to detect if a service supports regular OAuth 2.0 or service account (or both), sorry.

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+            {
+                var certificate = new X509Certificate2(keyFilePath, "notasecret", X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable);
+
+                ServiceAccountCredential credential = new ServiceAccountCredential(new ServiceAccountCredential.Initializer(serviceAccountEmail)
+                                                                                    {
+                                                                                        Scopes = scopes
+                                                                                    }.FromCertificate(certificate));
+
+                // Create the service.
+                var service = new AnalyticsService(new BaseClientService.Initializer()
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:192]: Just return here is enough.
return new AnalyticsService...

File: Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+                });
+                return service;
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.InnerException);
+                throw ex;
+            }
+        }
+		        /// <summary>
```
Comment (peleyal) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:205]: Please fix alignment :)

Comment (LindaLawton) [inline Analyticsv3Sample/AnalyticsSampleProject/StandardGoogleAuthSample.cs:205]: Alignment is fixed GitHub eats it when i upload it.  :(
